### PR TITLE
S2180

### PR DESCRIPTION
**Pull request by yeunga**
Added storage resolvers to caom2-artifact-resolvers for Story 2180.

### CODE REVIEW

File: caom2-artifact-resolvers/src/main/java/ca/nrc/cadc/caom2/artifact/resolvers/CadcMastResolver.java [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+ */
+public class CadcMastResolver implements StorageResolver {
+    public static final String SCHEME = "mast";
+    private static final Logger log = Logger.getLogger(CadcMastResolver.class);
+    private static final URI DATA_RESOURCE_ID = URI.create("ivo://cadc.nrc.ca/data");
+    private String baseDataURL;
+
+    public CadcMastResolver() {
+        try {
+            RegistryClient rc = new RegistryClient();
```
Comment (pdowler) [inline caom2-artifact-resolvers/src/main/java/ca/nrc/cadc/caom2/artifact/resolvers/CadcMastResolver.java]: delay until AuthMethod needed

File: caom2-artifact-resolvers/src/main/java/ca/nrc/cadc/caom2/artifact/resolvers/CadcMastResolver.java [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    }
+
+    @Override
+    public URL toURL(URI uri) {
+        if (!SCHEME.equals(uri.getScheme())) {
+            throw new IllegalArgumentException("invalid scheme in " + uri);
+        }
+
+        try {
+            URL url = new URL(this.baseDataURL + "/MAST/" + uri.getRawSchemeSpecificPart());
```
Comment (pdowler) [inline caom2-artifact-resolvers/src/main/java/ca/nrc/cadc/caom2/artifact/resolvers/CadcMastResolver.java]: why Raw and not getSchemeSpecificPart()?

File: caom2-artifact-resolvers/src/main/java/ca/nrc/cadc/caom2/artifact/resolvers/EsacMastResolver.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+/**
+ * StorageResolver implementation for the ESAC MAST archive.
+ * This class can convert an ESAC MAST URI into a URL. This is an alternate version that uses the RegistryClient to find the data web service base URL.
+ *
+ * @author yeunga
+ */
+public class EsacMastResolver implements StorageResolver {
+
+    public EsacMastResolver() {
+        throw new UnsupportedOperationException("");
```
Comment (pdowler) [inline caom2-artifact-resolvers/src/main/java/ca/nrc/cadc/caom2/artifact/resolvers/EsacMastResolver.java]: Constructor should be empty; that exception should be thrown from methods

File: caom2-artifact-resolvers/src/main/java/ca/nrc/cadc/caom2/artifact/resolvers/MastResolver.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+            newURL = MAST_BASE_ARTIFACT_URL + "/" + path;
+        } else {
+            log.error(CANNOT_GET_URL + " Invalid scheme (should be 'mast'): " + uri.toString());
+            throw new IllegalArgumentException(CANNOT_GET_URL + " Invalid scheme (should be 'mast'): " + uri.toString());
+        }
+
+        return newURL;
+    }
+
+    protected void validateScheme(URI uri) {
```
Comment (pdowler) [inline caom2-artifact-resolvers/src/main/java/ca/nrc/cadc/caom2/artifact/resolvers/MastResolver.java]: this should be called validate() as it does more than check scheme

and never log.error and throw

File: caom2-artifact-resolvers/src/main/java/ca/nrc/cadc/caom2/artifact/resolvers/GeminiResolver.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+        return sb.toString();
+    }
+
+    @Override
+    public String getScheme() {
+        return SCHEME;
+    }
+
+    protected void validateScheme(URI uri) {
```
Comment (pdowler) [inline caom2-artifact-resolvers/src/main/java/ca/nrc/cadc/caom2/artifact/resolvers/GeminiResolver.java]: this should be called validate() as it does more than check scheme

never log.error and throw - just throw and let the caller handle it

File: caom2-artifact-resolvers/src/test/java/ca/nrc/cadc/caom2/artifact/resolvers/AdResolverTest.java [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+
+    @Test
+    public void testFileHTTP() {
+        try {
+            adResolver.setAuthMethod(AuthMethod.ANON);
+            URI uri = new URI(FILE_URI);
+            URL url = adResolver.toURL(uri);
+            Assert.assertNotNull(url);
+            log.info("testFile: " + uri + " -> " + url);
+            Assert.assertEquals("http", url.getProtocol());
```
Comment (pdowler) [inline caom2-artifact-resolvers/src/test/java/ca/nrc/cadc/caom2/artifact/resolvers/AdResolverTest.java]: remove protocol check - this code is not responsible

File: caom2-artifact-resolvers/src/test/java/ca/nrc/cadc/caom2/artifact/resolvers/AdResolverTest.java [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+
+    @Test
+    public void testFileHTTPS() {
+        try {
+            adResolver.setAuthMethod(AuthMethod.CERT);
+            URI uri = new URI(FILE_URI);
+            URL url = adResolver.toURL(uri);
+            Assert.assertNotNull(url);
+            log.info("testFile: " + uri + " -> " + url);
+            Assert.assertEquals("https", url.getProtocol());
```
Comment (pdowler) [inline caom2-artifact-resolvers/src/test/java/ca/nrc/cadc/caom2/artifact/resolvers/AdResolverTest.java]: remove protocol check

File: caom2-artifact-resolvers/src/test/java/ca/nrc/cadc/caom2/artifact/resolvers/CadcMastResolverTest.java [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+    }
+
+    @Test
+    public void testToURL() {
+        try {
+            URI uri = new URI(FILE_URI);
+            URL url = cadcMastResolver.toURL(uri);
+            Assert.assertNotNull(url);
+            log.info("testFile: " + uri + " -> " + url);
+            Assert.assertEquals("http", url.getProtocol());
```
Comment (pdowler) [inline caom2-artifact-resolvers/src/test/java/ca/nrc/cadc/caom2/artifact/resolvers/CadcMastResolverTest.java]: remove protocol and path checks; they are part of intTest

File: caom2-artifact-resolvers/src/test/java/ca/nrc/cadc/caom2/artifact/resolvers/GeminiResolverTest.java [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+    }
+
+    @Test
+    public void testValidURI() {
+        try {
+            String uriStr = GeminiResolver.SCHEME + ":" + GeminiResolver.FILE_URI + "/" + VALID_FILE1;
+            URI uri = new URI(uriStr);
+            URL url = geminiResolver.toURL(uri);
+
+            Assert.assertEquals(PROTOCOL_STR, url.getProtocol());
```
Comment (pdowler) [inline caom2-artifact-resolvers/src/test/java/ca/nrc/cadc/caom2/artifact/resolvers/GeminiResolverTest.java]: remove url checks

File: caom2-artifact-resolvers/src/test/java/ca/nrc/cadc/caom2/artifact/resolvers/MastResolverTest.java [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+            List<String> validURIs = new ArrayList<String>();
+            validURIs.add(VALID_URI);
+            validURIs.add(VALID_URI2);
+
+            for (String uriStr : validURIs) {
+
+                URI uri = new URI(uriStr);
+                URL url = mastResolver.toURL(uri);
+
+                Assert.assertEquals(PROTOCOL_STR, url.getProtocol());
```
Comment (pdowler) [inline caom2-artifact-resolvers/src/test/java/ca/nrc/cadc/caom2/artifact/resolvers/MastResolverTest.java]: remove url checks

File: caom2-artifact-resolvers/src/test/java/ca/nrc/cadc/caom2/artifact/resolvers/VOSpaceResolverTest.java [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+
+    @Test
+    public void testFileHTTP() {
+        try {
+            vosResolver.setAuthMethod(AuthMethod.ANON);
+            URI uri = new URI(FILE_URI);
+            URL url = vosResolver.toURL(uri);
+            Assert.assertNotNull(url);
+            log.info("testFile: " + uri + " -> " + url);
+            Assert.assertEquals("http", url.getProtocol());
```
Comment (pdowler) [inline caom2-artifact-resolvers/src/test/java/ca/nrc/cadc/caom2/artifact/resolvers/VOSpaceResolverTest.java]: remove url checks

File: caom2-artifact-resolvers/src/test/java/ca/nrc/cadc/caom2/artifact/resolvers/VOSpaceResolverTest.java [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+
+    @Test
+    public void testFileHTTPS() {
+        try {
+            vosResolver.setAuthMethod(AuthMethod.CERT);
+            URI uri = new URI(FILE_URI);
+            URL url = vosResolver.toURL(uri);
+            Assert.assertNotNull(url);
+            log.info("testFile: " + uri + " -> " + url);
+            Assert.assertEquals("https", url.getProtocol());
```
Comment (pdowler) [inline caom2-artifact-resolvers/src/test/java/ca/nrc/cadc/caom2/artifact/resolvers/VOSpaceResolverTest.java]: ditto

File: caom2-artifact-resolvers/src/main/java/ca/nrc/cadc/caom2/artifact/resolvers/AdResolver.java [hunk 13]
```diff
@@ -122,6 +111,11 @@ public URL toURL(URI uri) {
 
         try {
             String path = getPath(uri);
+            
+            // check if authMethod has been set
+            if (this.authMethod == null) {
+                this.authMethod = AuthenticationUtil.getAuthMethod(AuthenticationUtil.getCurrentSubject());
```
Comment (pdowler) [inline caom2-artifact-resolvers/src/main/java/ca/nrc/cadc/caom2/artifact/resolvers/AdResolver.java]: Assigning the AuthMethod from the current subject to the member variable will make re-use of the resolver buggy; it should be assigned to the local var (am) instead.

File: caom2-artifact-resolvers/src/main/java/ca/nrc/cadc/caom2/artifact/resolvers/CadcMastResolver.java [hunk 14]
```diff
@@ -97,12 +97,12 @@
 
     public CadcMastResolver() {
         try {
-            RegistryClient rc = new RegistryClient();
             Subject subject = AuthenticationUtil.getCurrentSubject();
```
Comment (pdowler) [inline caom2-artifact-resolvers/src/main/java/ca/nrc/cadc/caom2/artifact/resolvers/CadcMastResolver.java]: I guess I misssed this yesterday: the ctor should not do any init at all. Both the AuthMethod and the base URL lookup should happen in the toURL method and no values kept in member variables. That way the impl will adapt to changes in the data web service capabilities (even if the resolver is kept for long period of time and re-used).

I would make the same change in AdResolver - no member variables (just static) and do the service lookup in the toURL method.

Comment (yeunga) [reply caom2-artifact-resolvers/src/main/java/ca/nrc/cadc/caom2/artifact/resolvers/CadcMastResolver.java]: So we don't support public void setAuthMethod(AuthMethod authMethod) in AdResolver?

**Pull request merged**